pkg/controller/instance/instance_controller.go

```diff
@@ -115,295 +76,201 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 					})
 				}
 			}
-			log.Printf("InstanceController: Found %v instances to reconcile for operator %v", len(requests), a.Meta.GetName())
 			return requests
 		})
 
-	// This map function makes sure that we *ONLY* handle created operatorVersion
-	ovEventFilter := predicate.Funcs{
-		CreateFunc: func(e event.CreateEvent) bool {
-			log.Printf("InstanceController: Received create event for: %v", e.Meta.GetName())
-			return true
-		},
-		UpdateFunc: func(e event.UpdateEvent) bool {
-			return false
-		},
-		DeleteFunc: func(e event.DeleteEvent) bool {
-			return false
-		},
-	}
-
-	if err = c.Watch(&source.Kind{Type: &kudov1alpha1.OperatorVersion{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: ovEventHandler}, ovEventFilter); err != nil {
-		return err
-	}
-
-	return nil
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&kudov1alpha1.Instance{}).
+		Owns(&kudov1alpha1.Instance{}).
+		Owns(&appsv1.Deployment{}).
+		Owns(&corev1.Service{}).
+		Owns(&batchv1.Job{}).
+		Owns(&appsv1.StatefulSet{}).
+		Watches(&source.Kind{Type: &kudov1alpha1.OperatorVersion{}}, &handler.EnqueueRequestsFromMapFunc{ToRequests: addOvRelatedInstancesToReconcile}).
+		Complete(r)
 }
 
-func instanceEventFilter(mgr manager.Manager) predicate.Funcs {
-	ctx := context.TODO()
-	return predicate.Funcs{
-		UpdateFunc: func(e event.UpdateEvent) bool {
-			old := e.ObjectOld.(*kudov1alpha1.Instance)
-			new := e.ObjectNew.(*kudov1alpha1.Instance)
-
-			// Get the OperatorVersion that corresponds to the new instance.
-			ov := &kudov1alpha1.OperatorVersion{}
-			err := mgr.GetClient().Get(ctx,
-				types.NamespacedName{
-					Name:      new.Spec.OperatorVersion.Name,
-					Namespace: new.GetOperatorVersionNamespace(),
-				},
-				ov)
-			if err != nil {
-				log.Printf("InstanceController: Error getting operatorversion \"%v\" for instance \"%v\": %v",
-					new.Spec.OperatorVersion.Name,
-					new.Name,
-					err)
-				// TODO: We probably want to handle this differently and mark this instance as unhealthy
-				// since it's linking to a bad OV.
-				return false
-			}
+// Reconcile is the main controller method that gets called every time something about the instance changes
+//
+//   +-------------------------------+
+//   | Query state of Instance       |
+//   | and OperatorVersion           |
+//   +-------------------------------+
+//                  |
+//                  v
+//   +-------------------------------+
+//   | Start new plan if required    |
+//   | and none is running           |
+//   +-------------------------------+
+//                  |
+//                  v
+//   +-------------------------------+
+//   | If there is plan in progress, |
+//   | proceed with the execution    |
+//   +-------------------------------+
+//                  |
+//                  v
+//   +-------------------------------+
+//   | Update instance with new      |
+//   | state of the execution        |
+//   +-------------------------------+
```

Comment:
❤️ 👏

Reply:
If that ain't a beautifly ASCII flow diagram, then I don't know what is! 👏